lib/batchingReceiver.ts

```diff
@@ -0,0 +1,143 @@
+      const finalAction = (timeOver: boolean, data?: EventData) => {
+        // Resetting the mode. Now anyone can call start() or receive() again.
+        this._receiver.removeListener(Constants.receiverError, onReceiveError);
+        this._receiver.removeListener(Constants.message, onReceiveMessage);
+        if (!data) {
+          data = eventDatas.length ? eventDatas[eventDatas.length - 1] : undefined;
+        }
+        if (!timeOver) {
+          clearTimeout(waitTimer);
+        }
+        if (this.receiverRuntimeMetricEnabled && data) {
+          this.runtimeInfo.lastSequenceNumber = data.lastSequenceNumber;
+          this.runtimeInfo.lastEnqueuedTimeUtc = data.lastEnqueuedTime;
+          this.runtimeInfo.lastEnqueuedOffset = data.lastEnqueuedOffset;
+          this.runtimeInfo.retrievalTime = data.retrievalTime;
+        }
+        return this.close().then(() => { resolve(eventDatas); });
+      };
+
+      // Action to be performed after the max wait time is over.
+      actionAfterWaitTimeout = () => {
+        timeOver = true;
+        finalAction(timeOver);
+      };
+
+      // Action to be performed on the "message" event.
+      onReceiveMessage = (context: rheaPromise.Context) => {
+        const data: EventData = EventData.fromAmqpMessage(context.message!);
+        if (eventDatas.length <= maxMessageCount) {
+          eventDatas.push(data);
+        }
+        if (eventDatas.length === maxMessageCount) {
+          finalAction(timeOver, data);
+        }
+      };
+
+      // Action to be taken when an error is received.
+      onReceiveError = (context: rheaPromise.Context) => {
+        this._receiver.removeListener(Constants.receiverError, onReceiveError);
+        this._receiver.removeListener(Constants.message, onReceiveMessage);
+        const error = translate(context.receiver.error);
+        debug("[%s] Receiver '%s' received an error:\n%O", this._context.connectionId, this.name, error);
+        if (waitTimer) {
+          clearTimeout(waitTimer);
+        }
+        reject(error);
+      };
+
+      const addCreditAndSetTimer = (reuse?: boolean) => {
+        debug("[%s] Receiver '%s', adding credit for receiving %d messages.", this._context.connectionId, this.name, maxMessageCount);
+        this._receiver.add_credit(maxMessageCount);
+        let msg: string = "[%s] Setting the wait timer for %d seconds for receiver '%s'.";
+        if (reuse) msg += " Receiver link already present, hence reusing it.";
+        debug(msg, this._context.connectionId, maxWaitTimeInSeconds, this.name);
+        waitTimer = setTimeout(actionAfterWaitTimeout, (maxWaitTimeInSeconds as number) * 1000);
+      };
+
+      if (!this._session && !this._receiver) {
+        debug("[%s] Receiver '%s', setting the prefetch count to 0.", this._context.connectionId, this.name);
+        this.prefetchCount = 0;
+        this._init(onReceiveMessage, onReceiveError).then(() => {
+          return addCreditAndSetTimer();
```

Comment:
nit: `() => addCreditAndSetTimer()` is equivalent to `() => { return addCreditAndSetTimer(); }`